src/index.js

```diff
@@ -0,0 +1,194 @@
+import { createElement, Component } from 'react';
+import generateOutsideCheck from './generateOutsideCheck';
+
+/**
+ * A higher-order-component for handling onClickOutside for React components.
+ */
+const registeredComponents = [];
+const handlers = [];
+
+/**
+ * This function generates the HOC function that you'll use
+ * in order to impart onOutsideClick listening to an
+ * arbitrary component. It gets called at the end of the
+ * bootstrapping code to yield an instance of the
+ * onClickOutsideHOC function defined inside setupHOC().
+ */
+export default function onClickOutsideHOC(WrappedComponent, config) {
+  return class onClickOutside extends Component {
+    static displayName = `OnClickOutside(${ WrappedComponent.displayName || WrappedComponent.name || 'Component' })`
+
+    static defaultProps = {
+      eventTypes: ['mousedown', 'touchstart'],
+      excludeScrollbar: (config && config.excludeScrollbar) || false,
+      outsideClickIgnoreClass: 'ignore-react-onclickoutside',
+      preventDefault: false,
+      stopPropagation: false,
+    }
+
+    static getClass = () => WrappedComponent.getClass ? WrappedComponent.getClass() : WrappedComponent
+
+    /**
+     * Access the WrappedComponent's instance.
+     */
+    getInstance() {
+      return WrappedComponent.prototype.isReactComponent ? this.instanceRef : this;
+    }
+
+    // this is given meaning in componentDidMount
+    __outsideClickHandler() {}
+
+    /**
+     * Add click listeners to the current document,
+     * linked to this component's state.
+     */
+    componentDidMount() {
+      // If we are in an environment without a DOM such
+      // as shallow rendering or snapshots then we exit
+      // early to prevent any unhandled errors being thrown.
+      if (typeof document === 'undefined' || !document.createElement){
+        return;
+      }
+
+      const instance = this.getInstance();
+      let clickOutsideHandler;
```

Comment:
aint sure how you judge what kind of scope do we have for a variable, i understand them all, but just quite dont get how you pick function's one here

whats the danger of `instance === this` here? im not that familiar with all implications of the codebase